File: renderers/web_core/src/v1_0/catalog/resolver.ts
/**
 * @fileoverview Strict 3-step Catalog Resolver for A2UI v1.0.
 *
 * Implements the 3-step catalog resolution fallback algorithm:
 * 1. Surface-specific catalog override
 * 2. Message-declared catalog
 * 3. Default catalog
 */

import {CatalogDefinition} from './catalog-definition.js';

export interface CatalogResolverOptions {
  /** Map of registered catalogs by catalogId */
  catalogs?: Map<string, CatalogDefinition> | Record<string, CatalogDefinition>;
  /** Default catalog ID to use as step 3 fallback */
  defaultCatalogId?: string;
}

export class CatalogResolver {
  private catalogsMap: Map<string, CatalogDefinition>;
  private defaultCatalogId?: string;

  constructor(options: CatalogResolverOptions = {}) {
    this.catalogsMap = new Map();
    if (options.catalogs) {
      if (options.catalogs instanceof Map) {
        options.catalogs.forEach((cat, id) => this.catalogsMap.set(id, cat));
      } else {
        Object.entries(options.catalogs).forEach(([id, cat]) => this.catalogsMap.set(id, cat));
      }
    }
    this.defaultCatalogId = options.defaultCatalogId;
  }

  /**
   * Registers or updates a catalog definition.
   */
  registerCatalog(catalog: CatalogDefinition): void {
    this.catalogsMap.set(catalog.catalogId, catalog);
  }

  /**
   * Resolves the effective catalog ID following the strict 3-step fallback:
   * 1. Surface-specific catalog override
   * 2. Message-declared catalog
   * 3. Default catalog ID
   */
  resolveCatalogId(surfaceOverrideId?: string, messageDeclaredId?: string): string | undefined {
    if (surfaceOverrideId && this.catalogsMap.has(surfaceOverrideId)) {
      return surfaceOverrideId;
    }
    if (messageDeclaredId && this.catalogsMap.has(messageDeclaredId)) {
      return messageDeclaredId;
    }
    if (this.defaultCatalogId && this.catalogsMap.has(this.defaultCatalogId)) {
      return this.defaultCatalogId;
    }
    // Return first match if given, even if not pre-registered in map
    return surfaceOverrideId || messageDeclaredId || this.defaultCatalogId;
  }

  /**
   * Resolves the CatalogDefinition using the 3-step fallback.
   */
  resolveCatalog(
    surfaceOverrideId?: string,
    messageDeclaredId?: string,
  ): CatalogDefinition | undefined {
    const catalogId = this.resolveCatalogId(surfaceOverrideId, messageDeclaredId);
    if (!catalogId) {
      return undefined;
    }
    return this.catalogsMap.get(catalogId);
  }
}

File: agent_sdks/python/a2ui_core/src/a2ui/core/catalog/resolver.py
"""Strict 3-step Catalog Resolver for A2UI v1.0.

Implements the 3-step catalog resolution fallback algorithm:
1. Surface-specific catalog override
2. Message-declared catalog
3. Default catalog
"""

from typing import Dict, Optional
from a2ui.core.catalog.catalog_definition import CatalogDefinition


class CatalogResolver:
    """Resolves effective catalog definitions using strict 3-step fallback."""

    def __init__(
        self,
        catalogs: Optional[Dict[str, CatalogDefinition]] = None,
        default_catalog_id: Optional[str] = None,
    ) -> None:
        """Initializes resolver with optional catalogs dictionary and default ID."""
        self._catalogs: Dict[str, CatalogDefinition] = (
            catalogs.copy() if catalogs else {}
        )
        self._default_catalog_id: Optional[str] = default_catalog_id

    def register_catalog(self, catalog: CatalogDefinition) -> None:
        """Registers a catalog definition."""
        self._catalogs[catalog.catalog_id] = catalog

    def resolve_catalog_id(
        self,
        surface_override_id: Optional[str] = None,
        message_declared_id: Optional[str] = None,
    ) -> Optional[str]:
        """Resolves the effective catalog ID following strict 3-step fallback.

        Step 1: Surface-specific catalog override
        Step 2: Message-declared catalog
        Step 3: Default catalog ID
        """
        if surface_override_id and surface_override_id in self._catalogs:
            return surface_override_id
        if message_declared_id and message_declared_id in self._catalogs:
            return message_declared_id
        if self._default_catalog_id and self._default_catalog_id in self._catalogs:
            return self._default_catalog_id

        # Return first non-None match even if not in catalogs map
        return surface_override_id or message_declared_id or self._default_catalog_id

    def resolve_catalog(
        self,
        surface_override_id: Optional[str] = None,
        message_declared_id: Optional[str] = None,
    ) -> Optional[CatalogDefinition]:
        """Resolves the CatalogDefinition using 3-step fallback."""
        catalog_id = self.resolve_catalog_id(surface_override_id, message_declared_id)
        if not catalog_id:
            return None
        return self._catalogs.get(catalog_id)
